### ISSUE TITLE
reportUnboundVariable - Assigned variable unbound after type narrowing using `pd.DataFrame`

### ISSUE THREAD

**Issue by sebas-intellegens** [microsoft/pyright#2540]
**Describe the bug**
The `Unbound` assertion of an assigned variable after type narrowing seems incorrect.
(Not sure how to best describe it)

Not sure if it's relevant but I noticed Pyright reports this as a possible unbound variable but Pylance (in VS Code) does not.

**To Reproduce**
Run pyright on this code snippet:

```python
from typing import Union
import pandas as pd

def fn_ub(param: Union[str, pd.DataFrame]):
    if isinstance(param, str):
        df = pd.DataFrame()
    elif isinstance(param, pd.DataFrame):
        df = param
    return df.shape # <-- Error: "df" is possibly unbound 
```

However this works

```python
from typing import List, Union

def fn(param: Union[str, List[str]]):
    if isinstance(param, str):
        l = []
    elif isinstance(param, list):
        l = param
    return len(l) # <-- OK: `l` has type `List[str]`
```

Does it have something to do with the `isinstance()` check on a non built-in type?
Is it expected that Pylance wouldn't complain about this? (not sure it's a relevant question)

**Expected behavior**
`df` has type `pd.DataFrame` and isn't considered possibly unbound.

**VS Code extension or command-line**
Are you running pyright as a VS Code extension or a command-line tool? Command-line tool
Which version? `pyright 1.1.184`

**Additional context**
Apologies if I missed this in the docs or issues (my search, https://github.com/microsoft/pyright/issues?q=is%3Aissue+unbound+after+assign, didn't return my specific error as far as I could see)

**Comment by erictraut** [comment]
The type checking logic is the same between pylance and pyright. There are two reasons you might be seeing a difference:
1. By default, all type checking diagnostics are disabled in pylance. The `python.analysis.typeCheckingMode` is "off". By contrast, pyright defaults to "basic" because most pyright users want type checking enabled.
2. Pylance ships with type stubs for some libraries including pandas. If you want to use these same type stubs with pyright, you can download them from [here](https://github.com/microsoft/python-type-stubs/tree/main/pandas). Create a directory in your project called `typings`, and place the `pandas` type stub directory within it.

If you do these things, you should see the same behavior between pylance and pyright.

The problem you're seeing with the `Unbound` diagnostic is evidence that you don't have the pandas type stubs installed. Type narrowing requires knowledge of types. Without the type stubs, `DataFrame` has an unknown type, so narrowing cannot be performed on `param`. That means that pyright can't determine whether you've exhaustively covered all cases in your `if/elif` statement, so it needs to assume that `df` may be unassigned (unbound to a value).

**Comment by erictraut** [comment]
I'm going to close this because I don't think there's a bug. If the above suggestions don't work for you, please let us know.

**Comment by sebas-intellegens** [comment]
> Pylance ships with type stubs for some libraries including pandas. If you want to use these same type stubs with pyright, you can download them from here. Create a directory in your project called typings, and place the pandas type stub directory within it.

This is what I was overlooking
Thanks for the explanation :)
